src/ssh_endpoint/main.go

```diff
@@ -0,0 +1,93 @@
+package ssh_endpoint
+
+import (
+	"fmt"
+	"log/slog"
+	"net/url"
+	"os/user"
+
+	"github.com/kevinburke/ssh_config"
+)
+
+type SshEndpoint struct {
+	Given string
+
+	GivenUser     string
+	SshConfigUser string
+	FallbackUser  string
+
+	GivenHost     string
+	SshConfigHost string
+
+	GivenPort     string
+	SshConfigPort string
+}
+
+func (e *SshEndpoint) String() string {
+	portPart := ""
+	if e.FinalPort() != "22" {
+		portPart = fmt.Sprintf(":%s", e.FinalPort())
+	}
+
+	return fmt.Sprintf("%s@%s%s", e.FinalUser(), e.FinalHost(), portPart)
+}
+
+func (e *SshEndpoint) FinalUser() string {
+	if e.GivenUser != "" {
+		return e.GivenUser
+	}
+
+	if e.SshConfigUser != "" {
+		return e.SshConfigUser
+	}
+
+	return e.FallbackUser
+}
+
+func (e *SshEndpoint) FinalHost() string {
+	if e.SshConfigHost != "" {
+		return e.SshConfigHost
+	}
+
+	return e.GivenHost
```

Comment:
Should comment why this one's a bit different than the rest when giving preference to config / given